service/storage_proxy.cc

```diff
@@ -3466,7 +3466,7 @@ storage_proxy::mutate_atomically_result(std::vector<mutation> mutations, db::con
                             auto local_addr = utils::fb_utilities::get_broadcast_address();
                             auto& topology = _tmptr->get_topology();
                             auto local_dc = topology.get_datacenter();
-                            auto& local_endpoints = topology.get_datacenter_racks().at(local_dc);
+                            auto local_endpoints = topology.get_datacenter_rack_endpoints(local_dc);
```

Comment:
It looks like proxy doesn't need the { rack -> [ endpoints ] } map here. It merely iterates over _all_ endpoints in the returned map and doesn't care which rack they belong to. With that, there's no need in getting the map for a given dc, we can go with only the topology.get_datacenter_endpoints() helper

Reply:
`endpoint_filter(gossiper, local_rack, local_endpoints);` below filters them further by `local_rack`

Reply:
@xemul `endpoint_filter(gossiper, local_rack, local_endpoints);` below does care about the local rack:
```c++
    if (validated.size() - validated.count(local_rack) >= 2) {
        // we have enough endpoints in other racks
        validated.erase(local_rack);
    }
```

Reply:
and change dropped in v2